pkg/controller/pxc/users_custom.go

```diff
@@ -0,0 +1,283 @@
+
+			err := um.Exec(ctx, alterUserQuery(&user, string(userSecret.Data[userSecretPassKey])))
+			if err != nil {
+				log.Error(err, "failed to update user", "user", user)
+				continue
+			}
+
+			err = k8s.AnnotateObject(ctx, r.client, userSecret,
+				map[string]string{annotationKey: sha256Hash(userSecret.Data[userSecretPassKey])})
+			if err != nil {
+				return errors.Wrap(err, "update user secret")
+			}
+
+			log.Info("User password updated", "user", user.Name)
+		}
+
+		us, err := um.GetUsers(ctx, user.Name)
+		if err != nil {
+			log.Error(err, "failed to get user", "user", user)
+			continue
+		}
+
+		if userChanged(us, &user) {
+			log.Info("Creating/updating user", "user", user.Name)
+
+			err := um.Exec(ctx, upsertUserQuery(&user, string(userSecret.Data[userSecretPassKey])))
+			if err != nil {
+				log.Error(err, "failed to update user", "user", user)
+				continue
+			}
+
+			err = k8s.AnnotateObject(ctx, r.client, userSecret,
+				map[string]string{annotationKey: sha256Hash(userSecret.Data[userSecretPassKey])})
+			if err != nil {
+				return errors.Wrap(err, "update user secret")
+			}
+
+			log.Info("User created/updated", "user", user.Name)
+		}
+	}
+
+	return nil
+}
+
+func generateUserPass(
+	ctx context.Context,
+	cl client.Client,
+	cr *api.PerconaXtraDBCluster,
+	secret *corev1.Secret,
+	passKey string) error {
+
+	log := logf.FromContext(ctx)
+
+	pass, err := generatePass()
+	if err != nil {
+		return errors.Wrap(err, "generate custom user password")
+	}
+
+	if secret.Data == nil {
+		secret.Data = make(map[string][]byte)
+	}
+	secret.Data[passKey] = pass
+
+	err = cl.Create(ctx, secret)
+	if err != nil {
+		return fmt.Errorf("create custom users secret: %v", err)
+	}
+
+	log.Info("Created custom user secrets", "secrets", cr.Spec.SecretsName)
+	return nil
+}
+
+func userPasswordChanged(secret *corev1.Secret, key, passKey string) bool {
+	if secret.Annotations == nil {
+		return false
+	}
+
+	hash, ok := secret.Annotations[key]
+	if !ok {
+		return false
+	}
+
+	newHash := sha256Hash(secret.Data[passKey])
+
+	return hash != newHash
+}
+
+func userChanged(current []users.User, new *api.User) bool {
+	if len(current) == 0 {
+		return true
+	}
+
+	if len(current) != len(new.Hosts) {
+		return true
+	}
+
+	newHosts := make(map[string]struct{}, len(new.Hosts))
+	for _, h := range new.Hosts {
+		newHosts[h] = struct{}{}
+	}
+
+	for _, u := range current {
+		if _, ok := newHosts[u.Host]; !ok {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getUserSecret gets secret by name defined by `user.PasswordSecretRef.Name` or creates
+// a default secret with defaultName
+func getUserSecret(ctx context.Context, cl client.Client, cr *api.PerconaXtraDBCluster, name, defaultName, passKey string) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	err := cl.Get(ctx, types.NamespacedName{Name: name, Namespace: cr.Namespace}, secret)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			if name == defaultName {
```

Comment:
It seems that we're returning an empty secret when the `name` parameter doesn't match `defaultName`.

```suggestion
		if k8serrors.IsNotFound(err) && name == defaultName {
```